src/oauth2/service.ts

```diff
@@ -169,6 +171,33 @@ export async function generateTokenFromCode(client: OAuth2Client, code: string):
 
 }
 
+export function validatePKCE(codeVerifier: string|undefined, codeChallenge: string|undefined, codeChallengeMethod: string|undefined) {
```

Comment:
Could we simply *not* call this function when PKCE is not set up? I feel a bit better about that than to make every argument potentially undefined. This means that this check has to move to the call-site

Reply:
I guess I can. Feels a bit weird to me to have something like this:
```
if (codeVerifier !! codeChallenge !! codeChallengeMethod) {
  validatePKCE(codeVerifier, codeChallenge, codeChallengeMethod)
}

Since PKCE is required by 2.1 and supported in 2.0, every token request can potentially require PKCE validation. It makes more logical sense to me for that to be the first thing to do while doing PKCE validation (do we even have to do PKCE validation). I don't like separating the "do we need to do PKCE validation" and the "do PKCE validation" since it's potentially applicable to every request.

Reply:
I think the risk of also allowing undefined for any of the parameters is that it gets easier to slip in bugs, and call that function with a bad argument.

If the function requires a string, the caller is forced to ensure that it really is a string and if it isn't, get a hard error. I feel that this makes the source a bit more resilient against potential bugs.

You might also consider passing in something else, like an object with all the properties. An issue here too is that given each of the arguments may or may not be undefined, this creates 8 different branches, whereas a type like this:

```
type PkceChallenge = {
  codeVerifier: string,
  codeChallenge: string,
  codeChallengeMethod: 'plain' | 'S256'
} | undefined
```
Only creates 2 branches, and it's a bit more obvious what the behavior of the 2 branches will be.

Reply:
Third alternative is to use overloads:

```typescript
 export function validatePKCE(codeVerifier: string, codeChallenge: string, codeChallengeMethod: string) 
 export function validatePKCE(codeVerifier: undefined, codeChallenge: undefined, codeChallengeMethod: undefined);
 export function validatePKCE(codeVerifier: string|undefined, codeChallenge: string|undefined, codeChallengeMethod: string|undefined) { 
```

This enforces that all arguments are either all undefined, or all strings.

Reply:
If I don't make the parameters optional or provide an object where the fields are optional, I can't do validation if the user does something wrong. 

For example, the user provides a challenge code in the authorization request but doesn't provide the verifier code in the token request (in which case `codeVerifier` would be missing). Or if the user provides the verified code in the token request, but didn't provide the challenge code + method in the authorization request (so `codeChallenge` and `codeChallengeMethod` would be missing). 

I could push all of that parameter validation up into `generateTokenFromCode`, but that's a lot of logic specific to doing PKCE validation inside a function meant to generate tokens.

Reply:
Fair, that sounds good!